### PR TITLE
Fix ConcurrentModificationException

### PR DESCRIPTION
**Pull request by jeremylong**
Initial fix for ConcurrentModificationException introduced by #3466.

### CODE REVIEW

File: core/src/main/java/org/owasp/dependencycheck/dependency/Dependency.java [hunk 1]
```diff
@@ -746,6 +746,13 @@ public synchronized void removeVulnerability(Vulnerability v) {
         return Collections.unmodifiableSet(relatedDependencies);
     }
 
+    /**
+     * Clears the {@link #relatedDependencies}.      *
```
Comment (jeremylong) [inline core/src/main/java/org/owasp/dependencycheck/dependency/Dependency.java]:
```suggestion
     * Clears the {@link #relatedDependencies}.
```

**Comment by aikebah** [comment]
Same fix is needed for DependencyMergingAnalyzer at lines [147-149](https://github.com/jeremylong/DependencyCheck/blob/6832c6434cf60df146f0ff6704c4c082cee13519/core/src/main/java/org/owasp/dependencycheck/analyzer/DependencyMergingAnalyzer.java#L147-L149).

**Comment by aikebah** [comment]
Sorry... overlooked that file in the changelog... It's already in there. So that leaves only FalsePostiveAnalyzer.removeBadMatches which removes items from vulnerableSoftwareIdentifiers while while iterating over them

```
exception: java.util.ConcurrentModificationException

java.util.TreeMap$PrivateEntryIterator.nextEntry(TreeMap.java:1211)
java.util.TreeMap$KeyIterator.next(TreeMap.java:1265)
java.util.Collections$UnmodifiableCollection$1.next(Collections.java:1044)
org.owasp.dependencycheck.analyzer.FalsePositiveAnalyzer.removeBadMatches(FalsePositiveAnalyzer.java:281)
```

**Review by aikebah — CHANGES_REQUESTED** [review]
These look good to me, and also get rid of a [potentially dangerous state-modifying peek](https://4comprehension.com/idiomatic-peeking/) that was signalled by SonarLint.

When a [FalsePositiveAnalyzer#removeBadMatches](https://github.com/jeremylong/DependencyCheck/blob/6832c6434cf60df146f0ff6704c4c082cee13519/core/src/main/java/org/owasp/dependencycheck/analyzer/FalsePositiveAnalyzer.java#L281) fix gets added as far as I can see no further ConcurrentModifcationExceptions should happen (all other uses of the Set-getters appear to be read-only usage, already employ a defensive copy or imlement a 'gather' phase followed by an 'update' phase in the code)

**Review by aikebah — DISMISSED** [review]
LGTM

File: core/src/main/java/org/owasp/dependencycheck/analyzer/FalsePositiveAnalyzer.java [hunk 2]
```diff
@@ -271,6 +271,7 @@ private void removeJreEntries(Dependency dependency) {
      */
     protected void removeBadMatches(Dependency dependency) {
 
+        Set<Identifier> toRemove = new HashSet<>();
```
Comment (jeremylong) [inline core/src/main/java/org/owasp/dependencycheck/analyzer/FalsePositiveAnalyzer.java]:
```suggestion
        final Set<Identifier> toRemove = new HashSet<>();
```

**Pull request merged**